TESTING.asciidoc

```diff
@@ -76,7 +76,177 @@ In order to start a node with a trial license execute the following command:
+           /      OS      \   
+          /      BWC       \   
+         / Internal Cluster \ 
+        /  External (REST)   \ 
+       /        Unit          \
+      /________________________\
+-------------------------------------------------
+
+As a rule of thumb, always prefer writing tests at the lowest possible level 
+of the pyramid. 
+
+==== Unit Tests 
+
+Unit tests are the preferred way to test some functionality: most of the time
+they are simpler to understand, more likely to reproduce, and unlikely to be
+affected by changes that are unrelated to the piece of functionality that is
+being tested.
+
+**`ESTestCase`**: The base class of all tests. 
+It is typically extended directly by unit tests.
+
+The Gradle target to run unit tests is **`test`**.
+
+==== External REST Tests 
+
+The `'elasticsearch.testclusters'` Gradle plugin has support for configuring
+and starting, and stopping clusters used for testing. 
+These test run against a cluster started from the generated distribution with 
+a configuration specific for the test and only interact using the REST APIs. 
+
+These can use the full default or oss distribution but we can also use a special
+one (integ-test) built for testing only that doesn't have any modules added.
+Module that are needed by the test can be added trough the testclusters plugin.
+This approach is preferred as it leads to better isolation, faster execution,
+and better build avoidance: if a change only changes a particular module, 
+all the others don't have to be re-tested, except when testing using the full 
+distribution.
+
+**`ESRestTestCase`**: is the base class for these tests.
+These tests can also be implemented using YAML and ran using a subclass of
+**ESClientYamlSuiteTestCase**.
+
+The Gradle target to run rest tests is **`integTest`**
+
+==== Internal Cluster Tests
+
+There are multiple base classes for these:
+
+* **`ESSingleNodeTestCase`**: This test case sets up a cluster that has a
+  single node.
+* **`ESIntegTestCase`**: An integration test case that creates a cluster that
+  might have multiple nodes.
+
+The reason why `ESSingleNodeTestCase` exists is that all our components used to
+be very hard to set up in isolation, which had led us to having a number of
+integration tests but close to no unit tests. `ESSingleNodeTestCase` is a
+workaround for this issue which provides an easy way to spin up a node and get
+access to components that are hard to instantiate like `IndicesService`.
+Whenever practical, you should prefer unit tests.
+
+Many tests extend `ESIntegTestCase`, mostly because this is how most tests used
+to work in the early days of Elasticsearch. However the complexity of these
+tests tends to make them hard to debug. Whenever the functionality that is
+being tested isn't intimately dependent on how Elasticsearch behaves as a
+cluster, it is recommended to write unit tests or REST tests instead.
+
+In short, most new functionality should come with unit tests, and optionally
```

Comment:
We need to come to some kind of agreement here. Internal cluster tests continue to be added to the codebase which somewhat conflicts with this statement. If we are going to encourage one thing vs the other we should have clearer guidelines here.